packages/react/scroll-area/src/ScrollArea.tsx

```diff
@@ -186,2234 +80,886 @@ type ScrollAreaPrimitive = Polymorphic.ForwardRefComponent<
-  Polymorphic.IntrinsicElement<typeof ScrollAreaCornerImpl>,
-  ScrollAreaCornerOwnProps
->;
+/* -----------------------------------------------------------------------------------------------*/
 
-const ScrollAreaCorner = React.forwardRef(function ScrollAreaCorner(props, forwardedRef) {
-  return useNativeScrollArea() ? null : <ScrollAreaCornerImpl {...props} ref={forwardedRef} />;
-}) as ScrollAreaCornerPrimitive;
+function toInt(value?: string) {
+  return value ? parseInt(value, 10) : 0;
+}
 
-ScrollAreaCorner.displayName = CORNER_NAME;
+function getThumbRatio(viewportSize: number, contentSize: number) {
+  const ratio = viewportSize / contentSize;
+  return isNaN(ratio) ? 0 : ratio;
+}
 
-/* ------------------------------------------------------------------------------------------------*/
+function getThumbSize(sizes: Sizes) {
+  const ratio = getThumbRatio(sizes.viewport, sizes.content);
+  const scrollbarPadding = sizes.scrollbar.paddingStart + sizes.scrollbar.paddingEnd;
+  const thumbSize = (sizes.scrollbar.size - scrollbarPadding) * ratio;
+  // minimum of 18 matches macOS minimum
+  return Math.max(thumbSize, 18);
+}
+
+function getScrollPositionFromPointer(
+  pointerPos: number,
+  pointerOffset: number,
+  sizes: Sizes,
+  dir: Direction = 'ltr'
+) {
+  const thumbSizePx = getThumbSize(sizes);
+  const thumbCenter = thumbSizePx / 2;
+  const offset = pointerOffset || thumbCenter;
+  const thumbOffsetFromEnd = thumbSizePx - offset;
+  const minPointerPos = sizes.scrollbar.paddingStart + offset;
+  const maxPointerPos = sizes.scrollbar.size - sizes.scrollbar.paddingEnd - thumbOffsetFromEnd;
+  const maxScrollPos = sizes.content - sizes.viewport;
+  const scrollRange = dir === 'ltr' ? [0, maxScrollPos] : [maxScrollPos * -1, 0];
+  const interpolate = linearScale([minPointerPos, maxPointerPos], scrollRange as [number, number]);
+  return interpolate(pointerPos);
+}
+
+function getThumbOffsetFromScroll(scrollPos: number, sizes: Sizes, dir: Direction = 'ltr') {
+  const thumbSizePx = getThumbSize(sizes);
+  const scrollbarPadding = sizes.scrollbar.paddingStart + sizes.scrollbar.paddingEnd;
+  const scrollbar = sizes.scrollbar.size - scrollbarPadding;
+  const maxScrollPos = sizes.content - sizes.viewport;
+  const maxThumbPos = scrollbar - thumbSizePx;
+  const scrollClampRange = dir === 'ltr' ? [0, maxScrollPos] : [maxScrollPos * -1, 0];
+  const scrollWithoutMomentum = clamp(scrollPos, scrollClampRange as [number, number]);
+  const interpolate = linearScale([0, maxScrollPos], [0, maxThumbPos]);
+  return interpolate(scrollWithoutMomentum);
+}
+
+function scrollingWithinScrollbarBounds(scrollPos: number, maxScrollPos: number) {
+  return scrollPos > 0 && scrollPos < maxScrollPos;
+}
+
+// Custom scroll handler to avoid scroll-linked effects
+// https://developer.mozilla.org/en-US/docs/Mozilla/Performance/Scroll-linked_effects
+const addUnlinkedScrollListener = (node: HTMLElement, handler = () => {}) => {
```

Comment:
Took me a bit of time to realised what you meant by "unlinkedScrollListener", mostly because the comment that talks about "scroll-linked effects" is down here with the implementation, but the first thing I saw when reading were the usage.

I think it would help to comment quite a bit around the usage in `ScrollAreaThumb`, as I'd say that's one of the more confusing part of the overall file.

Reply:
I'm adding a big comment here but is there a better name you can think of to rename this function? I wasn't sure what to call it 😬

Reply:
The comment works for me.